Feature/1690 add informative overlay

Mayisha opened the pull request:
Fixes #1690

### Description
The changes in this PR implements an informative overlay to provide users some information about Stripe. This overlay is only visible once the user enables UPE and visits `http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout` for the first time. This is a one time overlay and disappears forever when user click on the `Got it` button or `Cross` icon or outside of the modal.

# Testing instructions
- Enable UPE
- Visit `http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout`. The overlay should appear
- Dismiss the message. The overlay should disappear
- Refresh the page. The overlay should never appear again.

Mayisha committed:
Create InformationOverlay component

Mayisha committed:
Add placeholder and enque script to render the InformationOverlay

Mayisha committed:
Set priority

frosso reviewed (commented):
Would it be possible to have it match the designs closer before releasing?
It looks like the modal should appear next to the Stripe gateway, and the Stripe gateway should not be part of the overlay, possibly even with a blue border.
Playing with the z-index of the Stripe element should get us closer.

I also noticed that there's already a script that is being queued by the plugin on that page from another controller - maybe can leverage that instead of creating a new endpoint?

Mayisha committed:
Keep Stripe row within view

Mayisha committed:
highlight the Stripe row

Mayisha committed:
Enqueue overlay script inside WC_Stripe_Payment_Gateways_Controller class

Mayisha committed:
Refactor

Mayisha committed:
Add tests

Mayisha commented:
@frosso I have made some updates in the PR.
- Opening the modal closer to Stripe row, above or below depending on the Stripe row position. Also added a little scrolling there to make sure the Stripe row is always within the view
- Made the Stripe row out of the overlay. I have tried multiple ways there but could not make the border visible. So currently it is without the border (let me know if you have any solution). The border appears only if I make `position: absolute` but it mess up the other components.
- Using the existing controller to enqueue the script.
- Added test.

frosso commented on client/entrypoints/information-overlay/information-overlay.js:
```diff
@@ -0,0 +1,66 @@
+				'woocommerce-gateway-stripe'
+			) }
+			<Actions>
+				<Button onClick={ () => handleDismiss() } isPrimary>
```
```suggestion
				<Button onClick={ handleDismiss } isPrimary>
```

frosso commented on client/entrypoints/information-overlay/information-overlay.js:
```diff
@@ -0,0 +1,66 @@
+				'View your Stripe payment methods',
+				'woocommerce-gateway-stripe'
+			) }
+			onRequestClose={ () => handleDismiss() }
```
This way the reference to the `onRequestClose` function can stay the same
```suggestion
			onRequestClose={ handleDismiss }
```

frosso commented on webpack.config.js:
```diff
@@ -42,5 +42,7 @@ module.exports = {
 		upe_settings: './client/settings/index.js',
 		payment_gateways: './client/entrypoints/payment-gateways/index.js',
+		upe_information_overlay:
+			'./client/entrypoints/information-overlay/index.js',
```
Can either the entrypoint or the script be renamed so they more closely match?
```suggestion
			'./client/entrypoints/upe-information-overlay/index.js',
```

frosso commented on client/entrypoints/information-overlay/index.js:
```diff
@@ -0,0 +1,34 @@
+			: stripeRowTop;
+	// scrolling so that the Stripe row is always within view
+	jQuery( 'body,html' ).animate( { scrollTop }, 800, () => {
+		const informationOverlayContainer = document.getElementById(
+			'wc-stripe-information-overlay-container'
+		);
+
+		if ( informationOverlayContainer ) {
```
If the container is not present in the DOM, the Stripe row would be highlighted and the window would be scrolled.

What do you think about putting this check at the top?

(very minor) We could get rid of all this jQuery code and just use vanilla JS to assign the styles, in a `useEffect` on mount of the `InformationOverlay` component.

frosso commented on client/entrypoints/information-overlay/__tests__/index.test.js:
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import InformationOverlay from '../information-overlay';
```
The file being tested is `information-overlay`, can the file name on this test reflect that (instead of `index.test.js`)?

frosso reviewed (commented):
Great job, thanks for the changes!

I'm seeing some overlap in my setup 👀 can it be adjusted?
![Screen Shot 2021-10-05 at 7 48 28 AM](https://user-images.githubusercontent.com/273592/136025517-c10951ba-b871-402a-9a80-6573486ba8b7.png)
![Screen Shot 2021-10-05 at 7 51 18 AM](https://user-images.githubusercontent.com/273592/136025990-5c50b156-5436-4ed1-849e-307215946955.png)

Mayisha committed:
Rename files and component

Mayisha committed:
Fix scroll and modal position

Mayisha commented:
@frosso I have addressed the feedbacks. The overlapping thing was not happening in my local so it took me some time to reproduce. Please check if it is working now :innocent:

frosso commented on client/entrypoints/upe-information-overlay/index.js:
```diff
@@ -0,0 +1,26 @@
+			body.scrollTop = top;
+		}
+	}, 20 );
+}
```
There's something going on with this delayed mount, because when I first load the page I can't scroll anywhere, and the modal is not showing:
![2021-10-07 08 58 43](https://user-images.githubusercontent.com/273592/136399655-6628360e-d0a9-4c32-97c0-0c641824805a.gif)

Does this happen to you too?

Mayisha committed:
move scrolling logic inside the overlay component

Mayisha commented on client/entrypoints/upe-information-overlay/upe-information-overlay.js:
```diff
@@ -32,23 +32,43 @@ const UpeInformationOverlay = () => {
+		const top = body.scrollTop + 50;
+		if ( top > stripeRowTop ) {
+			clearInterval( scrollDown );
+			changeElementStyles();
```
Moved the scrolling inside the component itself. Changing the styles after the scroll is done. The modal may change it's position because of this.

frosso committed:
Merge branch 'develop' into feature/1690-add-informative-overlay
